### PR TITLE
XTGETTCAP+capabilities support

### PR DESCRIPTION
**Pull request by jquast**
Expand jinxed's virtual terminfo database and improve with runtime capabilities "overlay".

- docfixes: `README.rst` symlinked to `doc/intro.rst`, new file `doc/capabilities.rst`; 
- terminfo modules created using code generation for popular modern and legacy terminals
  - `codegen_terminfo.py` generates most of jinxed.terminfo module automatically from
    `terminals.txt` via new task, `tox -e codegen-terminfo`
  - added `_aliases.py` maps non-standard `TERM` values eg. `xterm-kitty` -> `kitty`, to their
    primary ncurses terminal capability name
- handwrote (ported/extracted) 'syncterm' and 'ansi-bbs', 'ansi'
- `Terminal.overlay_capabilities()` accepts runtime str/num/bool caps that take priority,
  enabling dynamic capability (designed for XTGETTCAP responses)

## Summary by Sourcery

Add a generated, ncurses-derived virtual terminfo database with alias resolution and runtime capability overlays, broaden terminal support, and update tooling, tests, and docs accordingly.

New Features:
- Expose the Terminal class from the top-level API and add Terminal.overlay_capabilities() to let callers override string, numeric, and boolean terminfo capabilities at runtime.
- Introduce a generated terminfo module set, including modern and legacy terminals (e.g. xterm-new variants, alacritty, kitty, ghostty, wezterm, rxvt*, linux, vt*, screen/tmux, ansi/iris-ansi, hpterm, putty, syncterm, ansi-bbs), plus an alias map so non-standard TERM values resolve to the correct module.
- Add a terminals.txt manifest and tests to ensure every listed and hand-maintained terminal module can be loaded successfully via setupterm().

Enhancements:
- Refactor existing xterm-derived terminfo modules to be based on xterm-new and synchronize them with the upstream ncurses database, including cleaning out delay tokens and charset designations and normalizing certain capabilities.
- Adjust Windows-focused terminfo overrides (e.g. ansicon, vtwin10) to align with the new terminfo generation style while preserving Windows-specific capability tweaks.
- Add stricter terminfo tests to enforce that certain charset capabilities are present-but-empty and that no STR_CAPS values contain G0/G1 designations or delay tokens.
- Introduce a code-generation script and tox environment to regenerate terminfo modules from the ncurses terminfo.src source while keeping the generated Python 2.7/3.x compatible.
- Update documentation configuration and intro/capabilities docs to describe the expanded terminal and capability support, and refresh copyright handling and Sphinx/RST tooling to cope with Sphinx-style roles.
- Advertise support for newer Python versions in setup metadata and adjust tox basepython usage so lint/docs/codegen run on a modern Python while still allowing legacy interpreters via an older host Python.

Documentation:
- Add and wire up new introductory and capabilities documentation pages, including a README symlink, to explain terminal coverage and capability behavior under the new terminfo system.

Tests:
- Extend terminal tests to cover overlay_capabilities() semantics, terminal alias resolution, and the ability to setupterm() for all terminals in terminals.txt plus hand-maintained modules.
- Enhance terminfo tests to validate structural invariants across all terminfo modules (including generated ones) and skip private helper modules during iteration.

Chores:
- Add funding metadata for GitHub Sponsors and include the ncurses license file required for distributing derived terminfo data.

### CODE REVIEW

File: doc/capabilities.rst [hunk 1]
```diff
@@ -0,0 +1,155 @@
+Capabilities
+============
+
+Jinxed provides a lightweight virtual `capabilities database` with its `terminfo(5)`_ family of
```
Comment (avylove) [inline doc/capabilities.rst]: Is capabilities database missing a trailing underscore?

File: doc/capabilities.rst [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  Jinxed provides no functions to manage historic terminal character sets and so they are removed.
+
+- The ``smacs``, ``rmacs``, ``enacs``, ``s0ds``, and ``s1ds`` capabilities are dropped entirely.
+
+  These enter/exit alternate character set operations switch between ASCII and line-drawing mode
+  via G0/G1 designation or Shift-Out/Shift-In.  Modern terminals use Unicode mapping (``acsc``)
+  for line-drawing characters instead.
+
+.. _blessed: https://pypi.org/project/blessed
```
Comment (avylove) [inline doc/capabilities.rst]: the blessed link doesn't seem to be used in this file

File: doc/faq.rst [hunk 3]
```diff
@@ -10,35 +10,41 @@
 Frequently Asked Questions
 ==========================
 
-
 Why is Jinxed called Jinxed?
 ----------------------------------
 
-Jinxed was written to support Blessed_ on Windows. We originally want to call it Cursed,
+Jinxed was written to support blessed_ on Windows. We originally want to call it Cursed,
```
Comment (avylove) [inline doc/faq.rst]: Not related to your changes, but since this line is being modified "want" should be "wanted"

File: doc/faq.rst [hunk 4]
```diff
@@ -10,35 +10,41 @@
...
----------------------------------------------------
+A select few terminal from `terminfo(5)`_ database were selected, excluding many historic and
+unlikely terminal types (like ``avatar``). Rather than distributed a full list of over 1,000
+code-generated python files, a small subset is selected.
 
-We are happy to add support for as many terminals as we can.
-However, not all terminals can be supported. There a few requirements.
+However, if we're missing support for a terminal please let us know its unique ``TERM`` name,
+whether it is in the `terminfo(5)`_ database, or provide an equivalent reference or directly `create
+an issue <https://github.com/Rockhopper-Technologies/jinxed/issues>`_ or pull request.
```
Comment (avylove) [inline doc/faq.rst]:
```suggestion
A few select terminals from the `terminfo(5)`_ database were included. Rather than distributed a full list of over 1,000 code-generated python files, many historic and unlikely terminal types (like ``avatar``) were excluded.

However, if we're missing support for a terminal please let us know. We will need its unique ``TERM`` name if it is in the `terminfo(5)`_ database or provide an equivalent reference. To submit, `create
an issue <https://github.com/Rockhopper-Technologies/jinxed/issues>`_ or, even better, a pull request.
```

File: doc/faq.rst [hunk 5]
```diff
@@ -10,35 +10,41 @@
...
 
-      We need to be able to identify the terminal in some reasonable way
-      and differentiate it from other terminals. This could be through environment variables,
-      the :py:mod:`platform` module, or some other method.
+1. The terminal must be detectable programmatically
 
-  2. Virtual terminal codes must be supported and documented
+   We need to be able to identify the terminal in some reasonable way
+   and differentiate it from other terminals. This could be through environment variables,
+   (eg. ``TERM``) the :py:mod:`platform` module, ``TTYPE`` telnet negotiation, or some other method.
```
Comment (avylove) [inline doc/faq.rst]: comma placement

```suggestion
   and differentiate it from other terminals. This could be through environment variables  (e.g. ``TERM``), the :py:mod:`platform` module, ``TTYPE`` telnet negotiation, or some other method.
```

File: doc/index.rst [hunk 6]
```diff
@@ -1,44 +1,17 @@
...
+=================================
+
 ..
   Copyright 2019 Avram Lubkin, All Rights Reserved
 
   This Source Code Form is subject to the terms of the Mozilla Public
   License, v. 2.0. If a copy of the MPL was not distributed with this
   file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-:github_url: https://github.com/Rockhopper-Technologies/jinxed
```
Comment (avylove) [inline doc/index.rst:8]: I think this is for the " Edit on GitHub" button on Read the Docs. Can also go in conf I think. Did it throw an error?

File: doc/spelling_wordlist.txt [hunk 7]
```diff
@@ -1,11 +1,19 @@
 Ansicon
 cbreak
+eg
```
Comment (avylove) [inline doc/spelling_wordlist.txt]: this is usually "e.g.", right? I think if we use "e.g." we can leave this out

File: doc/spelling_wordlist.txt [hunk 8]
```diff
@@ -1,11 +1,19 @@
...
+ncurses
+param
 programmatically
 reimplementation
+sgr
 stderr
 stdout
+termcap
 terminfo
+th
```
Comment (avylove) [inline doc/spelling_wordlist.txt]: This could cause mispellings of "the" to be missed. looks like it's only used for 9th, so can we just use ninth instead?

File: tox.ini [hunk 9]
```diff
@@ -15,7 +14,6 @@ envlist =
 
 
 [testenv]
-basepython = python3.13
```
Comment (avylove) [inline tox.ini:18]: Why move this to the individual environments?

File: jinxed/_terminal.py [hunk 10]
```diff
@@ -53,15 +54,58 @@ def __init__(self, term=None, fd=-1):  # pylint: disable=invalid-name
...
+        :arg dict str_caps: Mapping of capability name to string value.
+            String values are encoded as ``latin-1``; ``bytes`` values are
+            stored directly.  For non-ASCII text, prefer ``bytes`` to avoid
+            ambiguities with ``latin-1`` encoding.
+        :arg dict num_caps: Mapping of capability name to integer value.
+        :arg set bool_caps: Set of boolean capability names that are present.
+        """
+        if str_caps:
+            for k, v in str_caps.items():
+                self._overlay_str_caps[k] = v if isinstance(v, bytes) else v.encode('latin-1')
```
Comment (avylove) [inline jinxed/_terminal.py]: nit: key, val is preferred over single letter variables

**Review by avylove — COMMENTED** [review]
Looks awesome! I mostly had some proofreading and nits.

I would have dropped 3.5 and 2.7 before you started working on this if I realized they were still there, but let's do that in a different PR. This one is already big.

I didn't realize the workflow had stopped running. GitHub does that if a public repo doesn't have activity in a while. I re-enabled it, so hopefully the next push will kick it off, so we'll have to see what the tests look like.

File: tox.ini (hunk 9 above)
Comment (jquast) [reply tox.ini:18]: I need these changes to run python2.7 testing through tox, for primary python set to 3.6 with a tox+setuptools also compatible with launching python2.7, I needed to however tell the other steps to use a newer python

Anyway I have reverted everything but codegen, you can make any such changes as you wish

File: doc/index.rst (hunk 6 above)
Comment (jquast) [reply doc/index.rst:8]: This was rendering "plainly" in preview sphinx build, I will re-enable and link to it,

File: tox.ini (hunk 9 above)
Comment (jquast) [reply tox.ini:18]: I did have to include this CI change, 900e1fda645d05f141b54fa84c2a4d3ea713a871 to [fix py3.10](https://github.com/Rockhopper-Technologies/jinxed/actions/runs/25488209280/job/74789198242) and others from failing, `py3.10 - did you mean py310?`, pretty sure i have fixed this in other repos, that TOXENV used to be ignored when invalid and just run it whatever, so previously the default system 'python' version was used for versions not specified in the deleted conditional

File: doc/index.rst (hunk 6 above)
Comment (jquast) [reply doc/index.rst:8]: https://jinxed--8.org.readthedocs.build/en/8/

🤷‍♂

**Comment by jquast** [comment]
I ran an audit against collected XTGETTCAP's [from ucs-detect ](https://github.com/jquast/ucs-detect/) and found a few minor errors, will push some kind of "fixup" in the code generation and applied scripts

File: tox.ini (hunk 9 above)
Comment (avylove) [reply tox.ini:18]: Ok, do what you need to do and I'll play around with it once this is merged

File: doc/index.rst (hunk 6 above)
Comment (avylove) [reply doc/index.rst:8]: Oh well, it used to work. I guess we need to drop it

File: doc/faq.rst [hunk 11]
```diff
@@ -10,35 +10,40 @@
...
-Can you add support for _______ terminal?
----------------------------------------------------
+A select few terminals from the `terminfo(5)`_ database were selected to reduce packaging
+size, excluding many historic and unlikely terminal types like ``avatar``.
 
-We are happy to add support for as many terminals as we can.
-However, not all terminals can be supported. There a few requirements.
+If we are missing support for a terminal, please let us know its unique ``TERM`` name,
+whether it is in the `terminfo(5)`_ database, or provide an equivalent reference or directly
+`create an issue <https://github.com/Rockhopper-Technologies/jinxed/issues>`_ or pull request.
```
Comment (avylove) [inline doc/faq.rst]: I still think the wording here is hard to follow. Does this match what you intended?

```suggestion
A select few terminals from the `terminfo(5)`_ database were included to reduce packaging
size. Many historic and unlikely terminal types, like ``avatar``, were excluded.

If we are missing support for a terminal, please let us know its unique ``TERM`` name
and whether it is in the `terminfo(5)`_ database, or provide an equivalent reference. To submit,
`create an issue <https://github.com/Rockhopper-Technologies/jinxed/issues>`_
or `pull request <https://github.com/Rockhopper-Technologies/jinxed/pulls>`_.
```

File: doc/index.rst (hunk 6 above)
Comment (jquast) [reply doc/index.rst:8]: documentation says it should work, maybe it would when merged to master. In any case, I moved the option to `html_theme_options` it should work from there, also

File: codegen_terminfo.py [hunk 12]
```diff
@@ -0,0 +1,607 @@
...
+# termcap $HOME/.terminfo/g/ghostty, so we provide this alias for code generation on other systems.
+EXTRA_ALIASES = {'xterm-ghostty': 'ghostty', 'xterm-kitty': 'kitty'}
+
+GITHUB_BASE = 'https://github.com/Rockhopper-Technologies/jinxed/blob/main/jinxed/terminfo'
+
+# Fixups applied to capability data. These were discovered by comparing XTGETTCAP results form the
+# ucs-detect project https://github.com/jquast/ucs-detect/ vs. ncurses entries.
+
+
+def apply_fixups(data_map: dict[str, 'TermData']) -> None:
```
Comment (avylove) [inline codegen_terminfo.py:96]: These all seem to have the same pattern. Maybe down the road we can handle them programmaticly by combining the fixups and terminals.txt into more of a config file.

**Comment by jquast** [comment]
I've just added some microsoft terminals, pulling from latest terminfo.src, now distributing an entry for Windows Terminal.exe as [ms-terminal](https://invisible-island.net/ncurses/terminfo.src.html#tic-ms-terminal).

I found a distinction by environment variable of old conhost windows terminals vs. new "real terminal", and also `ANSICON` for the other bridging terminal [ansicon](https://github.com/adoxa/ansicon), in the next commit, matching the sources cited but I need to test them.

Edit: The only notable non-cosmetic difference is ka1, ka3, kc1, kc3, kent (home, pg_down, end, pg_up) are missing in ms-terminal but were present in vtwin10 -- however, I checked the patterns and they are all covered by [blessed.keyboard.DEFAULT_SEQUENCE_MIXIN](https://github.com/jquast/blessed/blob/5c80f358789548b5f2eb1975c340a1b3ac5bf2b7/blessed/keyboard.py#L2010) so it only otherwise grows or is cosmetically patched.

File: codegen_terminfo.py (hunk 12 above)
Comment (jquast) [reply codegen_terminfo.py:96]: I will try to do that now.

File: doc/index.rst (hunk 6 above)
Comment (jquast) [reply doc/index.rst:8]: I fixed it, it needs to be very near the top of the file, adding a heading caused it to be "too far" probably some `heading = lines[:20]` somewhere deep in sphinx there

**Comment by avylove** [comment]
This is looking awesome! What do you think of merging termnals.txt and terminals-fixups.toml into termnals.toml like this? We would need to use toml for parsing in older versions for test_terminals.py and add it to TESTS_REQUIRE in setup.py for <3.11, but that gets all the config information in one place and leaves the door open to other features we might want to add in the future, like additional aliases.

```toml
[alacritty]
[cons25]
[contour]
  patch = [
    # ncurses vt220+cvis: cnorm=\E[?25h                 (no mode 12)
    # contour Capabilities.cpp:104: "\033[?12l\033[?25h"
    { cap = "cnorm", from = "\\E[?25h", to = "\\E[?12l\\E[?25h" },
  ]
[domterm]
[dtterm]
[foot]
[ghostty]
[hpterm]
[iris-ansi]
[iTerm2.app]
[kitty]
  patch = [
    # ncurses att610+cvis0: cnorm=\E[?12l               (blink off)
    # kitty kitty/terminfo.py:128: 'cnorm': r'\E[?12h'  (blink on)
    { cap = "cnorm", from = "\\E[?12l\\E[?25h", to = "\\E[?12h\\E[?25h" },
  ]
[linux]
[linux-16color]
[mintty]
[mlterm]
[ms-terminal]
[pangoterm]
[putty]
[putty-256color]
[rio]
[rxvt]
[rxvt-256color]
[rxvt-unicode-256color]
[rxvt-unicode]
[screen]
[screen-256color]
[st]
[st-256color]
[sun]
[tmux]
[tmux-256color]
[vscode]
[vt100]
[vt102]
[vt220]
[vt320]
[warp]
[wezterm]
  patch = [
    # ncurses ansi+erase:  clear=\E[H\E[J               (no ED param)
    # wezterm wezterm.terminfo:40:    cnorm=\E[?12l\E[?25h, clear=\E[H\E[2J
    { cap = "cnorm", from = "\\E[?25h", to = "\\E[?12l\\E[?25h" },
    { cap = "clear", from = "\\E[H\\E[J", to = "\\E[H\\E[2J" },
    # ncurses xterm+256color2 uses colon separators (48:5: / 38:5:)
    # wezterm uses semicolons (48;5; / 38;5;)
    { cap = "setab", from = "\\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48:5:%p1%d%;m", to = "\\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m" },
    { cap = "setaf", from = "\\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38:5:%p1%d%;m", to = "\\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m" },
  ]
[winconsole]
[wy50]
[xterm]
[xterm-16color]
[xterm-256color]
[xterm-new]
```

**Review by avylove — APPROVED** [review]
Great work! Feel free to merge.

**Comment by jquast** [comment]
i'll release to pypi. thanks for review/!

**Comment by avylove** [comment]
It should not have been released as 1.x this is a major version release. Going to delete from pypi

**Comment by jquast** [comment]
We understand SEMVER differently, there was no breaking public API sorry for the trouble, the 1.5.0 was right in the PR

**Comment by avylove** [comment]
I think, while the end result should be seamless, the way the data is derived is different enough to warrant a major version bump. But no worries, it's all good now. Hope it didn't get in the way of your blessed PR work.
